### PR TITLE
Improve ncnn patch embed

### PR DESCRIPTION
**Pull request by tpoisonooo**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

* improve patch embed, remove redundant onnx shape opr

## Modification

* rewrite patch embed

before:
```mermaid
  flowchart  TD;
      conv-->shape;
      shape-->slice;
      slice-->concat;
      concat-->reshape;
      conv-->reshape;
```

after:
```mermaid
  flowchart  TD;
      conv-->reshape;
```

* if `get_onnx2ncnn_path` failed, give more tips

* fix-typo

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/mmcv/ops/transformer.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            (out_h, out_w).
+    """
+
+    if self.adaptive_padding:
+        x = self.adaptive_padding(x)
+
+    x = self.projection(x)
+    x_shape = x.shape
+    out_size = (x_shape[2], x_shape[3])
+    x = x.reshape((x_shape[0], x_shape[1], -1)).transpose(1, 2)
```
Comment (grimoire) [inline mmdeploy/mmcv/ops/transformer.py:29]: I think add a rewriter for `torch.Tensor.flatten` and `torch.flatten` would be more general.

File: mmdeploy/backend/ncnn/init_plugins.py [hunk 2]
```diff
@@ -24,7 +24,13 @@ def get_onnx2ncnn_path() -> str:
         str: A path of onnx2ncnn tool.
     """
     candidates = ['./onnx2ncnn', './onnx2ncnn.exe']
-    return get_file_path(os.path.dirname(__file__), candidates)
+    onnx2ncnn = get_file_path(os.path.dirname(__file__), candidates)
+    if not os.path.exists(onnx2ncnn):
+        onnx2ncnn = shutil.which('ncnn2int8')
```
Comment (grimoire) [inline mmdeploy/backend/ncnn/init_plugins.py:29]: 'ncnn2int8' -> 'onnx2ncnn' ?

**Comment by grimoire** [comment]
How to generate the graph in this PR? Looks cool.

**Comment by tpoisonooo** [comment]
> How to generate the graph in this PR? Looks cool.

Just markdown.

“ ```mermaid

”   flowchart  TD;

“       conv-->reshape;

” ```

File: mmdeploy/backend/ncnn/init_plugins.py (hunk 2 above)
Comment (tpoisonooo) [reply mmdeploy/backend/ncnn/init_plugins.py]: done.

File: mmdeploy/mmcv/ops/transformer.py (hunk 1 above)
Comment (tpoisonooo) [reply mmdeploy/mmcv/ops/transformer.py:29]: Rewriting `PosEmbed` is just to eliminate `shape` opr,  `torch.flatten` takes `start_axis` and `end_axis` into consideration. I don't plan to let the impact be too broad.

Comment (lvhan028) [reply mmdeploy/mmcv/ops/transformer.py:29]: I agree. If we change the behavior of `flatten`, even though it is feasible, it affects every model that uses `flatten` during the model conversion. And that's too much. Huge tests are needed.

**Review by lvhan028 — APPROVED** [review]

**Review by grimoire — APPROVED** [review]

**Pull request merged**